src/transformer/events/core/message_sent.php

```diff
@@ -0,0 +1,79 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Transformer fn for message sent event
+ *
+ * @package   logstore_xapi
+ * @copyright Daniel Bell <daniel@yetanalytics.com>
+ *            
+ * @license   https://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+namespace src\transformer\events\core;
+
+use src\transformer\utils as utils;
+use src\transformer\utils\get_activity as activity;
+
+/**
+ * Transformer fn for message sent event
+ *
+ * @param array $config The transformer config settings.
+ * @param \stdClass $event The event to be transformed.
+ * @return array
+ */
+
+function message_sent(array $config, \stdClass $event) {
+    global $CFG;
+    $repo = $config['repo'];
+    if (isset($event->objecttable) && isset($event->objectid)) {
+        $event_object = $repo->read_record_by_id($event->objecttable, $event->objectid);
+    } else {
+        $event_object = array();
+    }
+    
+    $user=$repo->read_record_by_id('user',$event->userid); 
+
+    $course = (isset($event->courseid) && $event->courseid != 0) ? $repo->read_record_by_id('course', $event->courseid) : null;
+
+    $lang = utils\get_course_lang(($course ? $course :  $repo->read_record_by_id('course',1)));
+    
+    $statement = [
+        'actor' => utils\get_user($config,$user),
+        'verb' => ['id' => 'http://activitystrea.ms/send',
+                   'display' => ['en' => 'Sent']],
+        'object' => [
+            'id' => $config['app_url'].'/course/view.php?id='.$event->objectid,
+            'definition' => [
+                'type' => "http://id.tincanapi.com/activitytype/chat-message",
+                'name' => [$lang => $event_object->subject ?? 'no subject'],
+                'description' => [$lang => $event_object->smallmessage],
+            ],
+        ],
+        'context' => [
+            'language' => $lang,
+            'contextActivities' => [
+                'category' => [activity\site($config)],
+            ],
+            'extensions' => utils\extensions\base($config, $event, $course)
```

Comment:
So I note the extensions are missing here. Originally these were special concepts from the EDLM profile but _should_ be prepositions. I'll get you an updated example file we must have missed these on the prepositions rundown.